cmd/promql-compliance-tester/main.go

```diff
@@ -15,17 +16,28 @@ import (
 	"github.com/promlabs/promql-compliance-tester/testcases"
 )
 
-func newPromAPI(url string) (v1.API, error) {
-	client, err := api.NewClient(api.Config{
-		Address: url,
-	})
+func newPromAPI(targetConfig config.TargetConfig) (v1.API, error) {
+	apiConfig := api.Config{Address: targetConfig.QueryURL}
+	if len(targetConfig.XQueryKey) > 0 {
+		apiConfig.RoundTripper = RoundTripperWithHeader{XQueryKey: targetConfig.XQueryKey}
+	}
+	client, err := api.NewClient(apiConfig)
 	if err != nil {
-		return nil, errors.Wrapf(err, "creating Prometheus API client for %q: %v", url, err)
+		return nil, errors.Wrapf(err, "creating Prometheus API client for %q: %v", targetConfig.QueryURL, err)
 	}
 
 	return v1.NewAPI(client), nil
 }
 
+type RoundTripperWithHeader struct {
+	XQueryKey string
+}
+
+func (rt RoundTripperWithHeader) RoundTrip(req *http.Request) (*http.Response, error) {
+	req.Header.Add("X-Query-Key", rt.XQueryKey)
```

Comment:
I had been looking at doing the same recently and saw that this would technically violate the [`RoundTripper` contract](https://golang.org/pkg/net/http/#RoundTripper):

```go
    // RoundTrip should not modify the request, except for
    // consuming and closing the Request's Body. RoundTrip may
    // read fields of the request in a separate goroutine. Callers
    // should not mutate or reuse the request until the Response's
    // Body has been closed.
```

I think this is ok for now, but it would be good to find a better solution long-term (unfortunately switching out the entire HTTP client in the Go Prometheus client would be more work), as it might lead to undefined behavior. At least we should add a `TODO` comment mentioning this.

Reply:
I'll add a comment.